datafusion/optimizer/src/type_coercion.rs

```diff
@@ -86,29 +87,45 @@ impl ExprRewriter for TypeCoercionRewriter {
     }
 
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
-        match expr {
+        match &expr {
             Expr::BinaryExpr { left, op, right } => {
                 let left_type = left.get_type(&self.schema)?;
                 let right_type = right.get_type(&self.schema)?;
-                let coerced_type = coerce_types(&left_type, &op, &right_type)?;
+                let coerced_type = coerce_types(&left_type, op, &right_type)?;
                 Ok(Expr::BinaryExpr {
-                    left: Box::new(left.cast_to(&coerced_type, &self.schema)?),
-                    op,
-                    right: Box::new(right.cast_to(&coerced_type, &self.schema)?),
+                    left: Box::new(
+                        left.as_ref().clone().cast_to(&coerced_type, &self.schema)?,
+                    ),
+                    op: *op,
+                    right: Box::new(
+                        right
+                            .as_ref()
+                            .clone()
+                            .cast_to(&coerced_type, &self.schema)?,
+                    ),
                 })
             }
+            Expr::Like { pattern, .. }
+            | Expr::ILike { pattern, .. }
+            | Expr::SimilarTo { pattern, .. } => match pattern.get_type(&self.schema)? {
```

Comment:
This seems like the wrong place to be type checking the argument to `Expr::Like`, etc. as the argument types to other exprs are checked so why would we treat `Expr::Like` differently?

I would expect that to be done in the SQL planner perhaps? or in the physical_expr conversion?

Reply:
We can't check the type of the `pattern` expression until we have a schema to resolve against (it could be a reference to a column or any other type of expression).

My goal was to do the validation in the logical plan for the benefit of other engines building on DataFusion that don't use the physical plan.

Reply:
Never mind, we do have the schema in SQL planning ... I will make that change.

Reply:
@alamb Ok, this PR just got a whole lot smaller. Maybe this really shouldn't have taken me 30 days to do :rofl: